packages/devtools_shared/lib/src/extensions/extension_manager.dart

```diff
@@ -142,27 +137,33 @@ class ExtensionsManager {
   }) async {
     _assertUriFormat(rootFileUriString);
     final List<Extension> extensions;
-    try {
-      // TODO(https://github.com/flutter/devtools/issues/7944): this assumes
-      // that the .dart_tool/package_config.json file is in the package root,
-      // which may be an incorrect assumption for pub workspaces.
-      final packageConfigPath = path.posix.join(
+    String packageConfigPath;
+    while (true) {
+      packageConfigPath = path.url.join(
         rootFileUriString,
         '.dart_tool',
         'package_config.json',
       );
-      extensions = await findExtensions(
-        'devtools',
-        packageConfig: Uri.parse(packageConfigPath),
-      );
-      logs.add(
-        'ExtensionsManager._addExtensionsForRoot find extensions for '
-        'config: $packageConfigPath, result: '
-        '${extensions.map((e) => e.package).toList()}',
-      );
-    } catch (e) {
-      rethrow;
+      if (File.fromUri(Uri.parse(packageConfigPath)).existsSync()) {
+        break;
+      }
+      final nextRootFileUriString = path.url.dirname(rootFileUriString);
+      if (nextRootFileUriString == rootFileUriString) {
+        // No package config was found.
+        return;
+      }
+      rootFileUriString = nextRootFileUriString;
     }
+    // TODO(sigurdm): Do we need to deduplicate package configs?
+    extensions = await findExtensions(
```

Comment:
can we put this logic in extension_discovery's findExtensions method instead? This same logic would apply to any type of extension, not just DevTools extensions so I think it should live there. I think we need to change findExtensions to take a path to a pubspec.yaml file instead of a package_config.json file, and then `findExtensions` can walk the directory structure to find the `package_config.json` file.

Reply:
Done